crates/store/src/state.rs

```diff
@@ -417,11 +456,114 @@ impl State {
             inner.blockchain.push(block_commitment);
         }
 
+        // After successful DB commit, query updated accounts' storage as well as vault data
+        // TODO look into making this consume the `account_tree_update`
+        self.update_storage_forest_from_db(updated_account_ids, block_num).await?;
+
         info!(%block_commitment, block_num = block_num.as_u32(), COMPONENT, "apply_block successful");
 
         Ok(())
     }
 
+    /// Updates `SmtForest` after a block is successfully applied
+    ///
+    /// Must be called after the DB transaction commits successfully, so we can safely
+    /// query the newly committed storage data.
+    ///
+    /// # Warning
+    ///
+    /// Has internal locking to mutate the state, use cautiously in scopes with other
+    /// mutex guards around!
+    ///
+    /// # Note
+    ///
+    /// The number of changed accounts is bounded by transactions per block.
+    async fn update_storage_forest_from_db(
+        &self,
+        changed_account_ids: Vec<AccountId>,
+        block_num: BlockNumber,
+    ) -> Result<(), ApplyBlockError> {
+        if changed_account_ids.is_empty() {
+            return Ok(());
+        }
+
+        self.update_storage_maps_in_forest(&changed_account_ids, block_num).await?;
+
+        self.update_vaults_in_forest(&changed_account_ids, block_num).await?;
+
+        Ok(())
+    }
```

Comment:
nit: I'm not sure "storage forest" is the right name here as the forest stores more than account storage data (i.e., vault data as well).

Also, the approach of writing the block, and only after that updating the `SmtForest` is probably not going to work in the medium/long term. Ideally, we should be updating the `SmtForest` during the `apply_block()` procedure. I haven't thought it through fully, but my intuition is that it would be both more efficient and require less code than what we have right now.

Reply:
> during the apply_block() procedure

I am not quite sure what you mean by that? Do you want to integrate it into the `AccountDelta` computation itself?